### PR TITLE
twister: pytest: Added fixture for mcumgr with unittests

### PR DESCRIPTION
**Pull request by gchwier**
Added mcumgr fixture to pytest-twister-harness. Added unittests for new fixture.

This is cherry-picked from PR #58393 
and there are examples, how new fixture can be used to test MCUBoot

### CODE REVIEW

**Comment by PerMac** [comment]
@de-nordic @nordicjm Maybe you could have a look if the commands for mcumgr usage have sense?

File: scripts/pylib/pytest-twister-harness/src/twister_harness/__init__.py
Comment (de-nordic) [inline scripts/pylib/pytest-twister-harness/src/twister_harness/__init__.py:10]:
```suggestion
__all__= ['Device', 'MCUmgr']
```

Comment (de-nordic) [reply scripts/pylib/pytest-twister-harness/src/twister_harness/__init__.py:10]: The name of the thing is MCUmgr.

File: scripts/pylib/pytest-twister-harness/src/twister_harness/fixtures/mcumgr.py
Comment (nordicjm) [inline scripts/pylib/pytest-twister-harness/src/twister_harness/fixtures/mcumgr.py:115]: *mcumgr (note all lowercase for go program, MCUmgr for protocol itself)

File: scripts/pylib/pytest-twister-harness/src/twister_harness/fixtures/mcumgr.py
Comment (gchwier) [reply scripts/pylib/pytest-twister-harness/src/twister_harness/fixtures/mcumgr.py:61]: This is just a default value and can be changed in tests, when calling `image_upload` .. but I know, that default values are not changes without special case.
What value could be good here, 30s?

Comment (nordicjm) [reply scripts/pylib/pytest-twister-harness/src/twister_harness/fixtures/mcumgr.py:61]: 30s is good, it might need extending for some boards e.g. QSPI but they're the exception, not the norm.

Comment (gchwier) [reply scripts/pylib/pytest-twister-harness/src/twister_harness/fixtures/mcumgr.py]: changed, thanks for comments

File: scripts/pylib/pytest-twister-harness/src/twister_harness/fixtures/mcumgr.py
Comment (gopiotr) [inline scripts/pylib/pytest-twister-harness/src/twister_harness/fixtures/mcumgr.py:56]: Please consider of using `check_output(shlex.split(command), text=True)` instead of `shell=True`, which is in general considered as potential source of vulnerabilities.
https://docs.python.org/3/library/subprocess.html#security-considerations
https://semgrep.dev/docs/cheat-sheets/python-command-injection/

Comment (gchwier) [reply scripts/pylib/pytest-twister-harness/src/twister_harness/fixtures/mcumgr.py]: Thanks for comment. I agree and changed to use `shlex` to avoid `shell=True`

**Review by gopiotr — DISMISSED** [review]

**Comment by hakehuang** [comment]
this looks like a debug like pyocd and jlink, just wonder why not add  to west support? and we can keep using the default device harness

**Comment by gchwier** [comment]
@hakehuang in our case it is used as a helper (pytest fixture) to automate using `mcumgr` tool. Only some method from MCUmgr are wrapped here, just methods required to upload second image, to test it and reset. Example usage of this fixture you can find here:
https://github.com/zephyrproject-rtos/zephyr/blob/d764f9ef4032d29577b79258b8de246f0c0e3d16/tests/boot/with_mcumgr/pytest/test_upgrade.py
(in draft PR #58393).
Sure, some support could be added to `west`, but then, for our case (using it in pytest scenarios) we need another wrapper for west-mcumgr commands (because we call `mcumgr` couple of times in one scenario, not just once like jlink for flashing).
If someone want to have MCUmgr wrapper in west, than we can open a separate issue (feature request) to west, but for our case it would be too much overhead.

**Comment by hakehuang** [comment]
> west-mcumgr commands (because we call `mcumgr` couple of times in one scenario, not just once like jlink for flashing).

west command can be extend according to your will, you can refer to west_commands/runner/jlink.py the do_add_parser function.
so you can put your board specific settings to your board.cmakefile for your mcumgr runner. and the twister framework can remains the same.
maybe we need think about it and discuss it in twg meeting. let me schedule it. could be defer to another issue.

**Review by hakehuang — DISMISSED** [review]

**Comment by nordicjm** [comment]
For the go mcumgr client, I am opposed to having it added as a west runner because the tool is unmaintained and buggy, there will be a replacement for it in the future, having the replacement added as a west runner would be a fine idea. Keeping the tool here used in tests allows for showing if code changes break it but it's not a tool to rely upon for new features.

**Comment by de-nordic** [comment]
> For the go mcumgr client, I am opposed to having it added as a west runner because the tool is unmaintained and buggy, there will be a replacement for it in the future, having the replacement added as a west runner would be a fine idea. Keeping the tool here used in tests allows for showing if code changes break it but it's not a tool to rely upon for new features.

I agree with the MCUmgr maintainer: the original MCUmgr-cli in go is no-go for the fixture.

**Review by nashif — CHANGES_REQUESTED** [review]
this needs some documenation and at least one user (a test) to demonstrate how it will be used.

**Comment by gchwier** [comment]
@nashif some examples to demonstrate how it will be used are in PR #58393 from where I cherry-picked this fixture.
I agree with adding some documentation about MCUmgr fixture to pytest-twister-harness section, and I will do that later

**Review by PerMac — APPROVED** [review]

**Review by gopiotr — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**
